File: ios/Packages/GenOSCore/Tests/GenOSCoreTests/TestSupport.swift
import Foundation
import Testing
@testable import GenOSCore

// MARK: - Manual clock

@MainActor
final class ManualClock: GenOSClock {
    private(set) var nowMs: Double = 0
    private var nextToken = 0
    private var scheduled: [(id: Int, at: Double, work: @MainActor () -> Void)] = []

    var now: Double { nowMs }

    final class Token: GenOSCancellable {
        let id: Int
        weak var clock: ManualClock?
        init(id: Int, clock: ManualClock) {
            self.id = id
            self.clock = clock
        }
        func cancel() { clock?.cancelToken(id) }
    }

    @discardableResult
    func schedule(afterMs: Double, _ work: @escaping @MainActor () -> Void) -> GenOSCancellable {
        nextToken += 1
        let id = nextToken
        scheduled.append((id: id, at: nowMs + afterMs, work: work))
        return Token(id: id, clock: self)
    }

    func cancelToken(_ id: Int) {
        scheduled.removeAll { $0.id == id }
    }

    /// Advance time, firing due callbacks in schedule order.
    func advance(by ms: Double) {
        let target = nowMs + ms
        while true {
            guard let next = scheduled.filter({ $0.at <= target }).min(by: { $0.at < $1.at }) else { break }
            scheduled.removeAll { $0.id == next.id }
            nowMs = max(nowMs, next.at)
            next.work()
        }
        nowMs = target
    }

    var pendingCount: Int { scheduled.count }
}

// MARK: - In-memory secure store with gated reads

actor GatedStorage {
    var values: [String: String] = [:]
    var readWaiters: [CheckedContinuation<Void, Never>] = []
    var gated = false
    var writeWaiters: [CheckedContinuation<Void, Never>] = []
    var writesGated = false
    var writeObservers: [@Sendable (String) -> Void] = []

    func setGated(_ g: Bool) { gated = g }
    func setWritesGated(_ g: Bool) { writesGated = g }
    func addWriteObserver(_ observe: @escaping @Sendable (String) -> Void) {
        writeObservers.append(observe)
    }
    func set(_ key: String, _ value: String?) {
        if let value { values[key] = value } else { values.removeValue(forKey: key) }
    }
    /// A write-path set: commits the value, then notifies write observers.
    func commitWrite(_ key: String, _ value: String?) {
        set(key, value)
        for observe in writeObservers { observe(key) }
    }
    func get(_ key: String) -> String? { values[key] }

    func waitIfGated() async {
        guard gated else { return }
        await withCheckedContinuation { readWaiters.append($0) }
    }

    func releaseReads() {
        gated = false
        let waiters = readWaiters
        readWaiters = []
        for w in waiters { w.resume() }
    }

    func waitIfWritesGated() async {
        guard writesGated else { return }
        await withCheckedContinuation { writeWaiters.append($0) }
    }

    func releaseWrites() {
        writesGated = false
        let waiters = writeWaiters
        writeWaiters = []
        for w in waiters { w.resume() }
    }
}

/// SecureStore test double. `gate()` blocks reads until `releaseReads()` so
/// tests can interleave user input with hydration; `gateWrites()` blocks
/// writes until `releaseWrites()` to simulate a hung keychain write.
struct MemorySecureStore: SecureStore {
    let storage = GatedStorage()

    func seed(_ key: String, _ value: String) async {
        await storage.set(key, value)
    }

    func gate() async { await storage.setGated(true) }
    func releaseReads() async { await storage.releaseReads() }
    func gateWrites() async { await storage.setWritesGated(true) }
    func releaseWrites() async { await storage.releaseWrites() }

    /// Write-observation seam: an AsyncStream that yields the key of each
    /// COMMITTED write. The stream buffers, so a fire-and-forget write that
    /// lands before the test's first `next()` is never lost - awaiting it is
    /// a deterministic replacement for polling on detached persists.
    func committedWrites() async -> AsyncStream<String> {
        let (stream, continuation) = AsyncStream.makeStream(of: String.self)
        await storage.addWriteObserver { continuation.yield($0) }
        return stream
    }

    func read(_ key: String) async throws -> String? {
        await storage.waitIfGated()
        return await storage.get(key)
    }

    func write(_ key: String, value: String?) async throws {
        await storage.waitIfWritesGated()
        await storage.commitWrite(key, value)
    }

    func stored(_ key: String) async -> String? {
        await storage.get(key)
    }
}

// MARK: - Scripted HTTP streaming

/// One scripted streaming response: a status plus SSE body byte chunks
/// delivered exactly as scripted (chunk boundaries preserved).
struct ScriptedResponse: Sendable {
    var status: Int = 200
    var chunks: [Data] = []
    /// Error thrown from the byte stream after the chunks (network failure).
    var streamErrorMessage: String? = nil
    /// Non-stream body for error statuses (res.text()).
    var errorBody: String = ""

    init(status: Int = 200, chunks: [Data] = [], streamErrorMessage: String? = nil, errorBody: String = "") {
        self.status = status
        self.chunks = chunks
        self.streamErrorMessage = streamErrorMessage
        self.errorBody = errorBody
    }

    static func sse(_ lines: [String], status: Int = 200) -> ScriptedResponse {
        ScriptedResponse(status: status, chunks: lines.map { Data(($0).utf8) })
    }
}

actor ScriptedHTTPState {
    var responses: [ScriptedResponse] = []
    var requests: [HTTPRequest] = []
    var chunksPulled = 0

    func push(_ r: ScriptedResponse) { responses.append(r) }
    func next(_ request: HTTPRequest) -> ScriptedResponse? {
        requests.append(request)
        guard !responses.isEmpty else { return nil }
        return responses.removeFirst()
    }
    func allRequests() -> [HTTPRequest] { requests }
    func recordPull() { chunksPulled += 1 }
    func pulled() -> Int { chunksPulled }
}

/// Pull-based chunk source: one chunk is handed out per consumer demand, and
/// Task cancellation is honored between chunks (the HTTPStreaming contract),
/// so tests can assert a cancelled stream loop stops pulling.
actor ScriptedChunkQueue {
    private var remaining: [Data]
    private let trailingErrorMessage: String?
    private let state: ScriptedHTTPState

    init(chunks: [Data], trailingErrorMessage: String?, state: ScriptedHTTPState) {
        remaining = chunks
        self.trailingErrorMessage = trailingErrorMessage
        self.state = state
    }

    func next() async throws -> Data? {
        try Task.checkCancellation()
        guard !remaining.isEmpty else {
            if let trailingErrorMessage { throw StreamError(trailingErrorMessage) }
            return nil
        }
        await state.recordPull()
        return remaining.removeFirst()
    }
}

struct ScriptedHTTP: HTTPStreaming, HTTPFetching {
    let state = ScriptedHTTPState()

    func enqueue(_ r: ScriptedResponse) async { await state.push(r) }

    func requests() async -> [HTTPRequest] { await state.allRequests() }

    /// Total SSE body chunks the consumer has pulled across all streams.
    func chunksPulled() async -> Int { await state.pulled() }

    /// Parsed JSON bodies of every request seen, oldest first.
    func requestBodies() async -> [JSONValue] {
        await state.allRequests().compactMap { req in
            req.body.flatMap { JSONValue.parse($0) }
        }
    }

    func stream(_ request: HTTPRequest) async throws -> (HTTPResponseHead, AsyncThrowingStream<Data, Error>) {
        guard let scripted = await state.next(request) else {
            throw StreamError("no scripted response")
        }
        let head = HTTPResponseHead(status: scripted.status)
        // For error statuses the RN reference reads the response body via
        // res.text(); surface errorBody through the byte stream so the
        // client's drain-and-decode path sees it.
        var chunks = scripted.chunks
        if !head.ok, chunks.isEmpty, !scripted.errorBody.isEmpty {
            chunks = [Data(scripted.errorBody.utf8)]
        }
        let queue = ScriptedChunkQueue(
            chunks: chunks,
            trailingErrorMessage: scripted.streamErrorMessage,
            state: state
        )
        let stream = AsyncThrowingStream<Data, Error>(unfolding: { try await queue.next() })
        return (head, stream)
    }

    func fetch(_ request: HTTPRequest) async throws -> (HTTPResponseHead, Data) {
        guard let scripted = await state.next(request) else {
            throw StreamError("no scripted response")
        }
        var body = Data()
        for c in scripted.chunks { body.append(c) }
        if body.isEmpty { body = Data(scripted.errorBody.utf8) }
        return (HTTPResponseHead(status: scripted.status), body)
    }
}

// MARK: - Fake screen streamer (controller tests)

/// Records every stream start; the test drives the handlers manually.
@MainActor
final class FakeStreamer: ScreenStreaming {
    struct Started {
        let messages: [ChatMessage]
        let handlers: StreamHandlers
        let token: StreamCancelToken
    }

    private(set) var started: [Started] = []

    func stream(messages: [ChatMessage], handlers: StreamHandlers, token: StreamCancelToken) {
        started.append(Started(messages: messages, handlers: handlers, token: token))
    }

    var last: Started? { started.last }
    var count: Int { started.count }
}

/// A ScreenStreaming impl that fires its handlers SYNCHRONOUSLY inside
/// stream() - regression double for the RN inflight-before-stream ordering
/// (a synchronous delta must not be dropped as stale).
@MainActor
final class SyncFiringStreamer: ScreenStreaming {
    var deltas: [String] = ["sync delta"]
    var finish = true

    func stream(messages: [ChatMessage], handlers: StreamHandlers, token: StreamCancelToken) {
        for d in deltas { handlers.onDelta(d) }
        if finish { handlers.onDone(StreamEndInfo(truncated: false, dropped: false)) }
    }
}

// MARK: - SSE payload builders

func sseContent(_ text: String) -> String {
    let escaped = text
        .replacingOccurrences(of: "\\", with: "\\\\")
        .replacingOccurrences(of: "\"", with: "\\\"")
        .replacingOccurrences(of: "\n", with: "\\n")
    return "data: {\"choices\":[{\"delta\":{\"content\":\"\(escaped)\"}}]}\n"
}

func sseFinish(_ reason: String) -> String {
    "data: {\"choices\":[{\"delta\":{},\"finish_reason\":\"\(reason)\"}]}\n"
}

let sseDone = "data: [DONE]\n"

/// Whole-call style tool call chunk (Cerebras): id+name+full arguments at once.
func sseWholeToolCall(index: Int, id: String, name: String, arguments: String) -> String {
    let escapedArgs = arguments
        .replacingOccurrences(of: "\\", with: "\\\\")
        .replacingOccurrences(of: "\"", with: "\\\"")
    return "data: {\"choices\":[{\"delta\":{\"tool_calls\":[{\"index\":\(index),\"id\":\"\(id)\",\"function\":{\"name\":\"\(name)\",\"arguments\":\"\(escapedArgs)\"}}]}}]}\n"
}

/// Split-arguments style fragment (OpenAI): only an arguments fragment.
func sseArgsFragment(index: Int, fragment: String) -> String {
    let escaped = fragment
        .replacingOccurrences(of: "\\", with: "\\\\")
        .replacingOccurrences(of: "\"", with: "\\\"")
    return "data: {\"choices\":[{\"delta\":{\"tool_calls\":[{\"index\":\(index),\"function\":{\"arguments\":\"\(escaped)\"}}]}}]}\n"
}

// MARK: - Stream test harness

/// Collected callback activity from one streamScreen run.
@MainActor
final class StreamRecorder {
    var deltas: [String] = []
    var doneInfos: [StreamEndInfo] = []
    var errors: [String] = []
    var toolRounds: [[ToolRoundCall]] = []
    var toolDecision: ToolRoundDecision = .proceed

    func handlers() -> StreamHandlers {
        StreamHandlers(
            onDelta: { [weak self] d in self?.deltas.append(d) },
            onDone: { [weak self] i in self?.doneInfos.append(i) },
            onError: { [weak self] e in
                self?.errors.append((e as? StreamError)?.message ?? String(describing: e))
            },
            onToolRound: { [weak self] calls in
                guard let self else { return .abort }
                self.toolRounds.append(calls)
                return self.toolDecision
            }
        )
    }

    var content: String { deltas.joined() }
}

/// A always-available fake tool executor with scripted outputs.
struct FakeTool: ToolExecuting, Sendable {
    var availableFlag: Bool = true
    var output: @Sendable (String, [String: JSONValue]) -> String = { name, args in
        "TOOL(\(name)):\(args["query"]?.stringValue ?? "")"
    }

    var available: Bool { availableFlag }
    var promptSection: String { "\n\n## Tools available" }
    var toolDefs: JSONValue {
        .array([.object([
            "type": .string("function"),
            "function": .object([
                "name": .string("web_search"),
                "description": .string("Search the live web."),
                "parameters": .object([
                    "type": .string("object"),
                    "properties": .object(["query": .object(["type": .string("string")])]),
                    "required": .array([.string("query")]),
                ]),
            ]),
        ])])
    }

    func execute(name: String, args: [String: JSONValue]) async -> String {
        output(name, args)
    }
}

@MainActor
func makeKeyStore(envKey: String? = "test-key", store: MemorySecureStore = MemorySecureStore()) -> KeyStore {
    KeyStore(envKey: envKey, store: store)
}

@MainActor
func makeStreamClient(
    http: ScriptedHTTP,
    keyStore: KeyStore? = nil,
    tools: ToolExecuting? = nil,
    systemPrompt: String = "SYSTEM",
    today: String = "Friday, July 25, 2026"
) -> StreamClient {
    StreamClient(
        http: http,
        keyStore: keyStore ?? makeKeyStore(),
        config: StreamConfig(systemPrompt: systemPrompt, todayString: { today }),
        tools: tools
    )
}

// MARK: - Controller harness

@MainActor
struct ControllerHarness {
    let clock = ManualClock()
    let streamer = FakeStreamer()
    let store: ScreenStore
    let controller: GenOSController

    init(apps: [AppDef] = []) {
        store = ScreenStore(clock: clock)
        controller = GenOSController(store: store, streamer: streamer, clock: clock, apps: apps)
    }

    /// Finish the most recent stream with plain content.
    func finishLast(content: String, truncated: Bool = false) {
        guard let s = streamer.last else { return }
        s.handlers.onDelta(content)
        s.handlers.onDone(StreamEndInfo(truncated: truncated, dropped: false))
    }
}

let sampleApp = AppDef(
    id: "weather",
    name: "Weather",
    emoji: "⛅",
    tileStart: "#5ac8fa",
    tileEnd: "#007aff",
    request: "Open the \"Weather\" app home screen."
)

File: android/genos-core/src/test/kotlin/dev/appless/genoscore/TestSupport.kt
package dev.appless.genoscore

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.test.TestScope
import kotlin.coroutines.coroutineContext

// Manual clock

/** Deterministic [GenOSClock]: time only moves when a test moves it. */
class ManualClock : GenOSClock {
    var nowMs: Double = 0.0
        private set

    private var nextToken = 0
    private val scheduled = mutableListOf<Entry>()

    private class Entry(val id: Int, val at: Double, val work: () -> Unit)

    override val now: Double
        get() = nowMs

    override fun schedule(afterMs: Double, work: () -> Unit): GenOSCancellable {
        val id = ++nextToken
        scheduled.add(Entry(id, nowMs + afterMs, work))
        return object : GenOSCancellable {
            override fun cancel() {
                scheduled.removeAll { it.id == id }
            }
        }
    }

    /** Advance time, firing due callbacks in schedule order. */
    fun advance(byMs: Double) {
        val target = nowMs + byMs
        while (true) {
            val next = scheduled.filter { it.at <= target }.minByOrNull { it.at } ?: break
            scheduled.removeAll { it.id == next.id }
            nowMs = maxOf(nowMs, next.at)
            next.work()
        }
        nowMs = target
    }

    /** Move the reading without firing anything (staleness tests). */
    fun jump(toMs: Double) {
        nowMs = toMs
    }

    val pendingCount: Int
        get() = scheduled.size
}

// In-memory secure store with gated reads/writes

/**
 * [SecureStore] test double. [gateReads] blocks reads until [releaseReads] so
 * tests can interleave user input with hydration; [gateWrites] blocks writes
 * until [releaseWrites] to simulate a hung keychain write. [failReads] makes the
 * read throw.
 */
class MemorySecureStore : SecureStore {
    val values = LinkedHashMap<String, String>()
    val writtenKeys = mutableListOf<String>()

    private var readGate: CompletableDeferred<Unit>? = null
    private var writeGate: CompletableDeferred<Unit>? = null
    var failReads: Boolean = false

    fun seed(key: String, value: String) {
        values[key] = value
    }

    fun gateReads() {
        readGate = CompletableDeferred()
    }

    fun releaseReads() {
        readGate?.complete(Unit)
        readGate = null
    }

    fun gateWrites() {
        writeGate = CompletableDeferred()
    }

    fun releaseWrites() {
        writeGate?.complete(Unit)
        writeGate = null
    }

    override suspend fun read(key: String): String? {
        readGate?.await()
        if (failReads) throw StreamException("keychain unavailable")
        return values[key]
    }

    override suspend fun write(key: String, value: String?) {
        writeGate?.await()
        writtenKeys.add(key)
        if (value == null) values.remove(key) else values[key] = value
    }
}

// Scripted HTTP

/**
 * One scripted streaming response: a status plus SSE body byte chunks delivered
 * exactly as scripted (chunk boundaries preserved).
 */
class ScriptedResponse(
    val status: Int = 200,
    val chunks: List<ByteArray> = emptyList(),
    /** Error thrown from the byte stream after the chunks (network failure). */
    val streamErrorMessage: String? = null,
    /** Non-stream body for error statuses (`res.text()`). */
    val errorBody: String = "",
) {
    companion object {
        fun sse(lines: List<String>, status: Int = 200): ScriptedResponse =
            ScriptedResponse(status = status, chunks = lines.map { it.toByteArray(Charsets.UTF_8) })

        fun bytes(chunks: List<ByteArray>, status: Int = 200): ScriptedResponse =
            ScriptedResponse(status = status, chunks = chunks)

        fun json(text: String, status: Int = 200): ScriptedResponse =
            ScriptedResponse(status = status, chunks = listOf(text.toByteArray(Charsets.UTF_8)))
    }
}

/**
 * Pull-based scripted transport. Chunks are handed out one per collector demand
 * and coroutine cancellation is honored between them, so tests can assert a
 * cancelled stream loop stops pulling. [gateAfterChunk] suspends the flow before
 * the Nth chunk until [releaseGate], giving deterministic mid-stream interleaving.
 */
class ScriptedHttp : HttpStreaming, HttpFetching {
    private val responses = ArrayDeque<ScriptedResponse>()
    val requests = mutableListOf<HttpRequest>()
    var chunksPulled: Int = 0
        private set

    var gateAfterChunk: Int? = null
    private var gate = CompletableDeferred<Unit>()

    fun releaseGate() {
        gate.complete(Unit)
    }

    fun enqueue(r: ScriptedResponse) {
        responses.addLast(r)
    }

    /** Parsed JSON bodies of every request seen, oldest first. */
    fun requestBodies(): List<JsonValue> = requests.mapNotNull { JsonValue.parse(it.bodyText) }

    override suspend fun stream(request: HttpRequest): Pair<HttpResponseHead, Flow<ByteArray>> {
        requests.add(request)
        val scripted = responses.removeFirstOrNull() ?: throw StreamException("no scripted response")
        val head = HttpResponseHead(scripted.status)
        // For error statuses the RN reference reads the body via res.text();
        // surface errorBody through the byte stream so the client's
        // drain-and-decode path sees it.
        val chunks = if (!head.ok && scripted.chunks.isEmpty() && scripted.errorBody.isNotEmpty()) {
            listOf(scripted.errorBody.toByteArray(Charsets.UTF_8))
        } else {
            scripted.chunks
        }
        val body = flow {
            chunks.forEachIndexed { i, chunk ->
                coroutineContext.ensureActive()
                gateAfterChunk?.let { if (i >= it) gate.await() }
                chunksPulled++
                emit(chunk)
            }
            scripted.streamErrorMessage?.let { throw StreamException(it) }
        }
        return head to body
    }

    override suspend fun fetch(request: HttpRequest): Pair<HttpResponseHead, ByteArray> {
        requests.add(request)
        val scripted = responses.removeFirstOrNull() ?: throw StreamException("no scripted response")
        var body = ByteArray(0)
        for (c in scripted.chunks) body += c
        if (body.isEmpty()) body = scripted.errorBody.toByteArray(Charsets.UTF_8)
        return HttpResponseHead(scripted.status) to body
    }
}

/** An [HttpFetching] whose fetch never returns (hung network). */
class HangingHttp : HttpFetching {
    val requests = mutableListOf<HttpRequest>()
    private val never = CompletableDeferred<Unit>()

    override suspend fun fetch(request: HttpRequest): Pair<HttpResponseHead, ByteArray> {
        requests.add(request)
        never.await()
        error("unreachable")
    }
}

// Fake screen streamer (controller tests)

/** Records every stream start; the test drives the handlers manually. */
class FakeStreamer : ScreenStreaming {
    class Started(
        val messages: List<ChatMessage>,
        val handlers: StreamHandlers,
        val token: StreamCancelToken,
    )

    val started = mutableListOf<Started>()

    override fun stream(
        messages: List<ChatMessage>,
        handlers: StreamHandlers,
        token: StreamCancelToken,
    ) {
        started.add(Started(messages, handlers, token))
    }

    val last: Started?
        get() = started.lastOrNull()

    val count: Int
        get() = started.size
}

/**
 * A [ScreenStreaming] impl that fires its handlers SYNCHRONOUSLY inside
 * `stream()` — the regression double for the RN inflight-before-stream ordering
 * (a synchronous delta must not be dropped as stale).
 */
class SyncFiringStreamer : ScreenStreaming {
    var deltas: List<String> = listOf("sync delta")
    var finish: Boolean = true

    override fun stream(
        messages: List<ChatMessage>,
        handlers: StreamHandlers,
        token: StreamCancelToken,
    ) {
        for (d in deltas) handlers.onDelta(d)
        if (finish) handlers.onDone(StreamEndInfo(truncated = false, dropped = false))
    }
}

// SSE payload builders

private fun jsonEscape(text: String): String = text
    .replace("\\", "\\\\")
    .replace("\"", "\\\"")
    .replace("\n", "\\n")

fun sseContent(text: String): String =
    "data: {\"choices\":[{\"delta\":{\"content\":\"${jsonEscape(text)}\"}}]}\n"

fun sseFinish(reason: String): String =
    "data: {\"choices\":[{\"delta\":{},\"finish_reason\":\"$reason\"}]}\n"

const val SSE_DONE: String = "data: [DONE]\n"

/** Whole-call style tool-call chunk (Cerebras): id + name + full arguments at once. */
fun sseWholeToolCall(index: Int, id: String, name: String, arguments: String): String =
    "data: {\"choices\":[{\"delta\":{\"tool_calls\":[{\"index\":$index,\"id\":\"$id\"," +
        "\"function\":{\"name\":\"$name\",\"arguments\":\"${jsonEscape(arguments)}\"}}]}}]}\n"

/** Split-arguments style fragment (OpenAI): only an `arguments` fragment. */
fun sseArgsFragment(index: Int, fragment: String): String =
    "data: {\"choices\":[{\"delta\":{\"tool_calls\":[{\"index\":$index," +
        "\"function\":{\"arguments\":\"${jsonEscape(fragment)}\"}}]}}]}\n"

// Stream test harness

/** Collected callback activity from one `streamScreen` run. */
class StreamRecorder {
    val deltas = mutableListOf<String>()
    val doneInfos = mutableListOf<StreamEndInfo>()
    val errors = mutableListOf<String>()
    val toolRounds = mutableListOf<List<ToolRoundCall>>()
    var toolDecision: ToolRoundDecision = ToolRoundDecision.PROCEED
    var onDeltaHook: ((String) -> Unit)? = null

    fun handlers(): StreamHandlers = StreamHandlers(
        onDelta = {
            deltas.add(it)
            onDeltaHook?.invoke(it)
        },
        onDone = { doneInfos.add(it) },
        onError = { errors.add((it as? StreamException)?.message ?: it.toString()) },
        onToolRound = {
            toolRounds.add(it)
            toolDecision
        },
    )

    val content: String
        get() = deltas.joinToString("")
}

/** An always-available fake tool executor with scripted outputs. */
class FakeTool(
    var availableFlag: Boolean = true,
    var output: (String, Map<String, JsonValue>) -> String = { name, args ->
        "TOOL($name):${args["query"]?.str ?: ""}"
    },
) : ToolExecuting {
    val executed = mutableListOf<Pair<String, Map<String, JsonValue>>>()

    override val available: Boolean
        get() = availableFlag

    override val promptSection: String
        get() = "\n\n## Tools available"

    override val toolDefs: JsonValue
        get() = JsonValue.Arr(
            listOf(
                JsonValue.obj(
                    "type" to JsonValue.Str("function"),
                    "function" to JsonValue.obj(
                        "name" to JsonValue.Str("web_search"),
                        "description" to JsonValue.Str("Search the live web."),
                        "parameters" to JsonValue.obj(
                            "type" to JsonValue.Str("object"),
                            "properties" to JsonValue.obj(
                                "query" to JsonValue.obj("type" to JsonValue.Str("string")),
                            ),
                            "required" to JsonValue.Arr(listOf(JsonValue.Str("query"))),
                        ),
                    ),
                ),
            ),
        )

    override suspend fun execute(name: String, args: Map<String, JsonValue>): String {
        executed.add(name to args)
        return output(name, args)
    }
}

fun makeKeyStore(
    envKey: String? = "test-key",
    store: MemorySecureStore = MemorySecureStore(),
    scope: CoroutineScope,
): KeyStore = KeyStore(envKey, store, scope)

fun makeStreamClient(
    http: ScriptedHttp,
    scope: CoroutineScope,
    keyStore: KeyStore? = null,
    tools: ToolExecuting? = null,
    systemPrompt: String = "SYSTEM",
    today: String = "Friday, July 25, 2026",
): StreamClient = StreamClient(
    http = http,
    keyStore = keyStore ?: makeKeyStore(scope = scope),
    config = StreamConfig(systemPrompt = systemPrompt, todayString = { today }),
    tools = tools,
    scope = scope,
)

// Controller harness

class ControllerHarness(apps: List<AppDef> = emptyList()) {
    val clock = ManualClock()
    val streamer = FakeStreamer()
    val store = ScreenStore(clock)
    val controller = GenOSController(store, streamer, clock, apps)

    /** Finish the most recent stream with plain content. */
    fun finishLast(content: String, truncated: Boolean = false) {
        val s = streamer.last ?: return
        s.handlers.onDelta(content)
        s.handlers.onDone(StreamEndInfo(truncated = truncated, dropped = false))
    }
}

val sampleApp: AppDef = AppDef(
    id = "weather",
    name = "Weather",
    emoji = "⛅",
    tileStart = "#5ac8fa",
    tileEnd = "#007aff",
    request = "Open the \"Weather\" app home screen.",
)

/**
 * A coroutine scope on the test dispatcher that `runTest` does NOT wait for.
 *
 * `runTest`'s own `backgroundScope` is deliberately excluded from
 * `advanceUntilIdle` (foreground work only), so a fire-and-forget persist or a
 * detached stream launched into it would never run inside an assertion window.
 * Reusing the test's context with a FRESH `Job` keeps the tasks foreground —
 * `advanceUntilIdle()` drains them — while detaching them from the test's job so
 * a deliberately hung coroutine cannot stall the test.
 */
fun TestScope.appScope(): CoroutineScope = CoroutineScope(coroutineContext + Job())
